ledger/eval/eval.go

```diff
@@ -1390,33 +1357,35 @@ func (eval *BlockEvaluator) endOfBlock() error {
 			return fmt.Errorf("fees collected wrong: %v != %v", eval.block.FeesCollected, expectedFeesCollected)
 		}
 
-		// agreement will check that the proposer is correct (we can't because
-		// we don't see the bundle), but agreement allows the proposer to be set
-		// even if Payouts is not enabled (and unset any time).  So make sure
-		// it's set iff it should be.
 		if eval.proto.Payouts.Enabled {
-			if eval.block.Proposer().IsZero() && !eval.generate { // if generating, proposer is set later by agreement
-				return fmt.Errorf("proposer missing when payouts enabled")
+			// agreement will check that the payout is zero if the proposer is
+			// ineligible, but we must check that it is correct if non-zero. We
+			// allow it to be too low. A proposer can be algruistic.
+			expectedPayout, err := eval.proposerPayout()
+			if err != nil {
+				return err
+			}
+			payout := eval.block.ProposerPayout()
+			if payout.Raw > expectedPayout.Raw {
```

Comment:
Any way different nodes could calculate different proposer payouts below the expected payout?

Reply:
`eval.proposerPayout()` should be deterministic. Validators are just saying "I don't care what the proposer put in _exactly_, just that it's not higher than allowed". This is similar to how timestamp works, where we don't check the value exactly, only that it follows the rules. (Most other things, we calculate an expected value and compare for equality.)